libsolidity/codegen/CompilerContext.cpp

```diff
@@ -188,57 +236,103 @@ bool CompilerContext::appendCallback(evmasm::AssemblyItem const& _i) {
+						}
+						// technically 0x44 is the minimum needed to add to length, but ABI wants right-padding so we overpad by 0x20.
+						kall(callBytes, add(0x64, length), callBytes, 0x20)
+						// kall to ovmREVERT will itself trigger safe reversion so nothing further needed! 
+					})",
+					{"length", "offset"});
+				break;
 			case Instruction::CREATE:
-				complexRewrite("ovmCREATE()", 3, 1, R"(
+				complexRewrite("ovmCREATE(bytes)", 3, 1, R"(
+						// methodId is stored for us at callBytes
+						let dataStart := add(callBytes, 4)
+						// store abi offset
+						mstore(dataStart, 0x20)
+						// store abi length
+						mstore(add(dataStart, 0x20), length)
+						// store bytecode itself
 						for { let ptr := 0 } lt(ptr, length) { ptr := add(ptr, 0x20) } {
-							mstore(add(add(callBytes, 4), ptr), mload(add(offset, ptr)))
+							mstore(add(add(dataStart, 0x40), ptr), mload(add(offset, ptr)))
 						}
-						let success := kall(callBytes, add(4, length), callBytes, 0x20)
-						if eq(success, 0) { revert(0, 0) }
+						// technically 0x44 is the minimum needed to add to length, but ABI wants right-padding so we overpad by 0x20.
+						kall(callBytes, add(0x64, length), callBytes, 0x20)
+						// legnth is first stack val in ==> first stack val out (address)
 						length := mload(callBytes)
+
+						// remove all the stuff we did at callbytes.
+						let newMemSize := msize()
+						for { let ptr := callBytes } lt(ptr, newMemSize) { ptr := add(ptr, 0x20) } {
+							mstore(add(callBytes, ptr), 0x00)
+						}
 					})",
 					{"length", "offset", "value"});
 				break;
 			case Instruction::CREATE2:
-				complexRewrite("ovmCREATE2()", 4, 1, R"(
-						mstore(add(callBytes, 4), salt)
+				complexRewrite("ovmCREATE2(bytes,bytes32)", 4, 1, R"(
+						// methodId is stored for us at callBytes
+						let dataStart := add(callBytes, 4)
+						// store abi offset
+						mstore(dataStart, 0x40)
+						// store salt
+						mstore(add(dataStart, 0x20), salt)
+						// store abi length
+						mstore(add(dataStart, 0x40), length)
+						// store bytecode itself
 						for { let ptr := 0 } lt(ptr, length) { ptr := add(ptr, 0x20) } {
-							mstore(add(add(callBytes, 0x24), ptr), mload(add(offset, ptr)))
+							mstore(add(add(dataStart, 0x60), ptr), mload(add(offset, ptr)))
 						}
-						let success := kall(callBytes, add(0x24, length), callBytes, 0x20)
-						if eq(success, 0) { revert(0, 0) }
+						// technically 0x64 is the minimum needed to add to length, but ABI wants right-padding so we overpad by 0x20.
+						kall(callBytes, add(0x84, length), callBytes, 0x20)
+						// salt is first stack val in ==> first stack val out (address)
 						salt := mload(callBytes)
+
+						// remove all the stuff we did at callbytes.
+						let newMemSize := msize()
+						for { let ptr := callBytes } lt(ptr, newMemSize) { ptr := add(ptr, 0x20) } {
+							mstore(add(callBytes, ptr), 0x00)
```

Comment:
Shouldn't this just be `mstore(ptr, 0x00)`

Reply:
good catch